experimental/LieAlgebras/src/LieSubalgebra.jl

```diff
@@ -0,0 +1,296 @@
+end
+
+###############################################################################
+#
+#   Constructions
+#
+###############################################################################
+
+@doc raw"""
+    bracket(S1::LieSubalgebra, S2::LieSubalgebra) -> LieAlgebraIdeal
+
+Return $[S_1, S_2]$.
+"""
+function bracket(
+  S1::LieSubalgebra{C,LieT}, S2::LieSubalgebra{C,LieT}
+) where {C<:RingElement,LieT<:LieAlgebraElem{C}}
+  @req base_lie_algebra(S1) === base_lie_algebra(S2) "Incompatible Lie algebras."
+  return ideal(base_lie_algebra(S1), [x * y for x in gens(S1) for y in gens(S2)])
+end
+
+###############################################################################
+#
+#   Important ideals and subalgebras
+#
+###############################################################################
+
+@doc raw"""
+    normalizer(L::LieAlgebra, S::LieSubalgebra) -> LieSubalgebra
+
+Return the normalizer of `S` in `L`, i.e. $\{x \in L \mid [x, S] \subseteq S\}$.
+"""
+function normalizer(L::LieAlgebra, S::LieSubalgebra)
+  @req base_lie_algebra(S) == L "Incompatible Lie algebras."
+
+  mat = zero_matrix(coefficient_ring(L), dim(L) + dim(S)^2, dim(L) * dim(S))
+  for (i, bi) in enumerate(basis(L))
+    for (j, sj) in enumerate(basis(S))
+      mat[i, ((j - 1) * dim(L) + 1):(j * dim(L))] = _matrix(bracket(bi, sj))
+    end
+  end
+  for i in 1:dim(S)
+    mat[(dim(L) + (i - 1) * dim(S) + 1):(dim(L) + i * dim(S)), ((i - 1) * dim(L) + 1):(i * dim(L))] = basis_matrix(
+      S
+    )
+  end
+  sol_dim, sol = left_kernel(mat)
+  sol = sol[:, 1:dim(L)]
+  c_dim, c_basis = rref(sol)
+  return sub(L, [L(c_basis[i, :]) for i in 1:c_dim]; is_basis=true)
+end
+
+function normalizer(S::LieSubalgebra)
+  return normalizer(base_lie_algebra(S), S)
+end
+
+@doc raw"""
+    centralizer(L::LieAlgebra, S::LieSubalgebra) -> LieSubalgebra
+  
+Return the centralizer of `S` in `L`, i.e. $\{x \in L \mid [x, S] = 0\}$.
+"""
+function centralizer(L::LieAlgebra, S::LieSubalgebra)
+  return centralizer(L, basis(S))
+end
+
+function centralizer(S::LieSubalgebra)
+  return centralizer(base_lie_algebra(S), basis(S))
+end
+
+###############################################################################
+#
+#   Properties
+#
+###############################################################################
+
+@doc raw"""
+    is_self_normalizing(S::LieSubalgebra) -> Bool
+
+Return `true` if `S` is self-normalizing, i.e. if its normalizer is `S`.
+"""
+function is_self_normalizing(S::LieSubalgebra)
+  return normalizer(base_lie_algebra(S), S) == S
+end
+
+###############################################################################
+#
+#   Conversion
+#
+###############################################################################
+
+@doc raw"""
+    lie_algebra(S::LieSubalgebra) -> LieAlgebra
+
+Return `S` as a Lie algebra.
+"""
+function lie_algebra(S::LieSubalgebra)
+  return lie_algebra(basis(S)) #, embedding_hom   # TODO
```

Comment:
There are other places, too, where an independent new algebraic structure gets created from a given substructure. Either the embedding morphism should be returned by the function that constructs the new object, or a function `is_subalgebra` (or so) should be provided that computes the embedding on demand (similar to the situation with `is_subfield` or `is_subgroup`).

Reply:
As reflected by the comment in the code, this is planned for a future PR. As one would first introduce a new type of morphisms, this seemed to be too involved for this PR, but will be added in the future.